proposals/0000-wire-protocol.md

```diff
@@ -0,0 +1,435 @@
+
+Let's consider an example. Suppose we want to fetch block 0, and we have 2 and 3 but not 5. We need to therefore communicate:
+
+```
+the leading bit is a parent, not an uncle
+we already have the first uncle, 2 so don't send us that
+we don't have the next uncle, 5
+we have the next parent, 3
+```
+
+We would encode this into the bit vector `1011`. Decoded:
+
+```
+101(1) <-- the leading bit is a parent, not an uncle
+10(1)1 <-- we already have the first uncle, 2 so don't send us that
+1(0)11 <-- we don't have the next uncle, 5
+(1)000 <-- we have the next parent, 3
+```
+      
+So using this digest the recipient can easily figure out that they only need to send us one hash, 5, for us to verify block 0.
+
+The bit vector 1 (only contains a single one) means that we already have all the hashes we need so just send us the block.
+
+These digests are very compact in size, only `(log2(number-of-blocks) + 2) / 8` bytes needed in the worst case. For example if you are sharing one trillion blocks of data the digest would be `(log2(1000000000000) + 2) / 8 ~= 6` bytes long.
+
+
+# Examples
+
+## Simple Download
+
+Alice has an e-book identified by a public-key (PK) which Bob would like to download. The e-book also has a discovery-key (DK). Bob connects to Alice, and the following exchange occurs:
+
+```
+  BOB: sends Feed          {discoveryKey: DK, nonce: BobNonce}
+  BOB: sends Handshake     {id: BobID}
+ALICE: sends Feed          {discoveryKey: DK, nonce: AliceNonce}
+ALICE: sends Handshake     {id: AliceID}
+  BOB: waits for Feed 
+  BOB: waits for Handshake
+ALICE: waits for Feed
+ALICE: waits for Handshake
+  BOB: sends Want          {start: 0}
+ALICE: receives Want       {start: 0}
+ALICE: sends Have          {start: 0, bitfield: ...}
+  BOB: receives Have       {start: 0, bitfield: ...}
+  BOB: sends Request       (for block 0)
+ALICE: receives Request    (for block 0)
+ALICE: sends Data          (for block 0)
+  BOB: receives Data       (for block 0)
+  BOB: sends Request       (for block 1)
+ALICE: receives Request    (for block 1)
+ALICE: sends Data          (for block 1)
+   ... (repeat for all other blocks) ...
+  BOB: sends Info          {downloading: false}
+ALICE: closes connection
+```
+
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Encryption might not make sense in some contexts (eg, IPC, or if the transport layer is already providing encryption). Should this DEP recognize this explicitly?
+- What parts of the design do you expect to resolve through the DEP consensus process before this gets merged?
+- What parts of the design do you expect to resolve through implementation and code review, or are left to independent library or application developers?
+- What related issues do you consider out of scope for this DEP that could be addressed in the future independently of the solution that comes out of this DEP?
+- How do "ack"s work?
```

Comment:
I'm curious also!

Reply:
From grepping around hypercore/hypercore-protocol code, as far as I can tell `ack` doesn't actually do anything, other than getting passed around as an config flag.
I'm guessing that what it is intended to do is have a peer reply with a new `Have` message in response to every `Data` message it receives? @mafintosh can you clarify?